
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Pseudobam</title>
    <meta name="description" content="">
    <meta name="author" content="Sreeram Kannan">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <!-- Le HTML5 shim, for IE6-8 support of HTML elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le styles -->
    <link href="/assets/themes/twitter/bootstrap/css/bootstrap.2.2.2.min.css" rel="stylesheet">
    <link href="/assets/themes/twitter/css/style.css?body=1" rel="stylesheet" type="text/css" media="all">

    <!-- Le fav and touch icons -->
  <!-- Update these with your own images
    <link rel="shortcut icon" href="images/favicon.ico">
    <link rel="apple-touch-icon" href="images/apple-touch-icon.png">
    <link rel="apple-touch-icon" sizes="72x72" href="images/apple-touch-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="114x114" href="images/apple-touch-icon-114x114.png">
  -->

    <!-- atom & rss feed -->
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">

  </head>

  <body>
    <div class="navbar">
      <div class="navbar-inner">
        <div class="container-narrow">
          <a class="brand" href="/">Shannon</a>
          <ul class="nav">
            
            
            


  
    
      
      	
      	<li><a href="/about.html">About</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/download.html">Download</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/faq.html">FAQ</a></li>
      	
      
    
  
    
  
    
      
      	
      	<li><a href="/kallisto_about.html">About</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/kallisto_download.html">Download</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/kallisto_faq.html">FAQ</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/license.html">License</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/manual.html">Manual</a></li>
      	
      
    
  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/source.html">Source</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/starting.html">Getting Started</a></li>
      	
      
    
  



          </ul>
        </div>
      </div>
    </div>

    <div class="container-narrow">

      <div class="content">
        
<div class="page-header">
  <h1>Pseudobam </h1>
</div>

<div class="row-fluid">
  <div class="span12">
    
<p>The output of <code>kallisto quant</code> when run with the <code>--pseudobam</code> option is a SAM formatted stream. This stream can be redirected to a SAM file or directly to a BAM file by piping into <code>samtools view -Sb - &gt; output.bam</code>. Examples of how to do this are shown in the <a href="manual.html">Manual</a>.</p>

<p>The SAM/BAM format is described in <a href="http://samtools.github.io/hts-specs/SAMv1.pdf">http://samtools.github.io/hts-specs/SAMv1.pdf</a></p>

<p><code>kallisto</code> will output a SAM file showing the pseudoalignments of the reads to the transcriptome. Each read will have one SAM entry for each transcript is pseudoaligns to.</p>

<p>For paired end reads the output will first include all entries for the first read, followed by all entries for the second read.</p>

<p>Each SAM line, where a read pseudoaligns to a transcript, follows the normal SAM format with the following interpretation of the fields</p>

<ul>
  <li><code>QNAME</code>, read name, any /1 or /2 is discarded from read names and the information is stored in the FLAG field</li>
  <li><code>FLAG</code>, normal SAM flags</li>
  <li><code>RNAME</code>, the transcript name</li>
  <li><code>POS</code>, 1-based position of the leftmost mapping</li>
  <li><code>MAPQ</code>, 255 for pseudoaligned reads, 0 otherwise</li>
  <li><code>CIGAR</code>, all matches (probably a lie), e.g. for 75bp reads <code>75M</code></li>
  <li><code>RNEXT</code>, the transcript of the mate, always equal to <code>RNAME</code></li>
  <li><code>PNEXT</code>, the leftmost 1-based position of the mate</li>
  <li><code>TLEN</code>, observed template length</li>
  <li><code>SEQ</code>, read sequence or the reverse complement if it pseudoaligns to the reverse complement of the transcript</li>
  <li><code>QUAL</code>, the quality values of the read</li>
  <li>optional fields
    <ul>
      <li><code>NH</code>, set to the size of the equivalence class, i.e. the number of transcripts this reads pseudoaligns to.</li>
    </ul>
  </li>
</ul>

<p>Example output (skipping header lines)</p>

<pre><code>ENST00000623973_197_729_1:0:0_2:0:0_8c  99      ENST00000623973 197     255     70M     659     =       532     CCTGGTGACCATCCTTGACTCCCGCCTGCACACGCCCAAGTACTTCTTCCTAGGGAACCTCTCCTTCCTG  2222222222222222222222222222222222222222222222222222222222222222222222  NH:i:3
ENST00000623973_197_729_1:0:0_2:0:0_8c  99      ENST00000341959 197     255     70M     659     =       532     CCTGGTGACCATCCTTGACTCCCGCCTGCACACGCCCAAGTACTTCTTCCTAGGGAACCTCTCCTTCCTG  2222222222222222222222222222222222222222222222222222222222222222222222  NH:i:3
ENST00000623973_197_729_1:0:0_2:0:0_8c  99      ENST00000625188 141     255     70M     603     =       532     CCTGGTGACCATCCTTGACTCCCGCCTGCACACGCCCAAGTACTTCTTCCTAGGGAACCTCTCCTTCCTG  2222222222222222222222222222222222222222222222222222222222222222222222  NH:i:3
ENST00000623973_197_729_1:0:0_2:0:0_8c  147     ENST00000623973 659     255     70M     197     =       -532    GTGACGAATGTGATCTTCCTAGGAGTCCCGGTTCTGTTCTTCTCTTTCTCCTATGTCTTCAGCATCACCA  2222222222222222222222222222222222222222222222222222222222222222222222  NH:i:3
ENST00000623973_197_729_1:0:0_2:0:0_8c  147     ENST00000341959 659     255     70M     197     =       -532    GTGACGAATGTGATCTTCCTAGGAGTCCCGGTTCTGTTCTTCTCTTTCTCCTATGTCTTCAGCATCACCA  2222222222222222222222222222222222222222222222222222222222222222222222  NH:i:3
ENST00000623973_197_729_1:0:0_2:0:0_8c  147     ENST00000625188 603     255     70M     141     =       -532    GTGACGAATGTGATCTTCCTAGGAGTCCCGGTTCTGTTCTTCTCTTTCTCCTATGTCTTCAGCATCACCA  2222222222222222222222222222222222222222222222222222222222222222222222  NH:i:3
</code></pre>

<p>In this case the read <code>ENST00000623973_197_729_1:0:0_2:0:0_8c</code>, which was simulated from <code>ENST00000623973</code> was aligned to 3 transcripts, one of which is <code>ENST00000623973</code> and the mapping positions match the read name. To decode the SAM flags consult <a href="https://broadinstitute.github.io/picard/explain-flags.html">this handy guide</a>.</p>

  </div>
</div>


      </div>
      <hr>
      <footer>
        <p>&copy; 2015 Sreeram Kannan
          with help from <a href="http://jekyllbootstrap.com" target="_blank" title="The Definitive Jekyll Blogging Framework">Jekyll Bootstrap</a>
          and <a href="http://twitter.github.com/bootstrap/" target="_blank">Twitter Bootstrap</a>
        </p>
      </footer>

    </div>

    


  </body>
</html>

